### ISSUE TITLE
Possible training speedups

### ISSUE THREAD

**Issue by josephrocca** [nicklashansen/tdmpc2#18]
In the last few days I've been playing around trying to see how fast I can get a 19M model training on a single 4090. My somewhat arbitrary goal is 1 hour, down from about 24 hours (just on `humanoid-walk` for now). I was pretty surprised at how well simple `torch.mean` weight-merging worked (y-axis is reward):

![image](https://github.com/nicklashansen/tdmpc2/assets/1167575/b81e6f39-17cb-4024-86ba-68818708939d)

The yellow line is the "control" run - i.e. no sharing/merging, and the bunched lines are 'workers' (separate Python processes) which share weights via the filesystem every so often. It seems like 4 workers with merging every 2.5k steps works quite well. The above chart shows 8 workers, but it only very marginally improves on 4 workers in my tests.

(Side note: Kind of counter-intuitive how 4 workers seem to give >4x speedup, but increasing beyond that barely improves speed. Might need to run some more tests to confirm - I've only tried 8 workers once, whereas I've tried 4 workers in several experiments.)

In case anyone is interested, the main code for this (in `online_trainer.py` while loop) is just:

<details>
<summary><b>Click to show code</b></summary>

```py
# be sure to change the seed for each worker
if cfg.do_multitrain_model_merging and self._step >= num_seed_steps and self._step > 0 and self._step % cfg.model_merge_freq == 0 and cfg.model_merge_index != 0: # zeroth model is the 'control' model
  model_path = f'/mnt/tdmpc2/tdmpc2/temp_worker_models/{cfg.exp_name}-{cfg.model_merge_index}.pt'
  lock_path = f'{model_path}.lock'
  with FileLock(lock_path, timeout=200): # prevent others from reading the model file while we are writing to it
    self.agent.save(model_path)
  agents = []
  for filename in os.listdir('/mnt/tdmpc2/tdmpc2/temp_worker_models/'):
    # check regex match for filename `{cfg.exp_name}-[0-9]+.pt`
    regex_match = f'^{cfg.exp_name}-[0-9]+\\.pt$'
    if re.match(regex_match, filename):
      file_path = os.path.join('/mnt/tdmpc2/tdmpc2/temp_worker_models/', filename)
      lock_path = f'{file_path}.lock'
      # wait to acquire lock before reading the model file
      with FileLock(lock_path, timeout=200):
        print("Loading model:", filename)
        a = TDMPC2(cfg)
        a.load(file_path)
        agents.append(a)
  # merge all the models into the current agent's model
  model_names = ["_encoder", "_dynamics", "_reward", "_pi", "_Qs", "_target_Qs"]
  for model_name in model_names:
    print("merging model_name:", model_name)
    models = [getattr(a.model, model_name) for a in agents]
    for model in models:
      model.eval()
    with torch.no_grad():
      for name, param in getattr(self.agent.model, model_name).named_parameters():
        params_to_merge = [model.state_dict()[name] for model in models]
        mean_param = torch.mean(torch.stack(params_to_merge, dim=0), dim=0)
        param.data.copy_(mean_param)
```
</details>

I'm quite happy with a ~5x speedup for such a small change to the code, and am now on the hunt for another 5x to get it under an hour.

I'm wondering if Nick or anyone else here has any thoughts on some more low-hanging fruit? I thought perhaps that a generalized reward component that effectively aims to teach the agent "inverse kinematics" in the very early stages might be helpful (basically giving it desired global coordinates for random body parts), and I've started experimenting with ways to do that, but no luck so far.

Any loose thoughts/tips/ideas would be useful - I'm a newbie and haven't yet got great intuitions for trimming the experiment search space.

**Comment by nicklashansen** [comment]
This looks really interesting! I'll have to think about this a bit more, but just to clarify: does each worker only perform model updates or are they entirely separate training processes? I.e., does each worker collect their own data? I am planning to release an implementation soon that trades sample-efficiency with better wall-time using environment parallelization (stepping in multiple copies of the environment in parallel). I imagine that env parallelization and training parallelization like you're seemingly proposing here could be used simultaneously.

**Comment by nicklashansen** [comment]
This all makes a lot of sense @josephrocca! I'd be very interested to see where this leads to :-)

> *Very* keen to try that out!

I now have it up and running internally but still need to run some performance benchmarks. I will let you know as soon as it is ~ready for external testing!

**Comment by nicklashansen** [comment]
@josephrocca I have pushed my implementation to branch [vectorized_env](https://github.com/nicklashansen/tdmpc2/tree/vectorized_env). I verified that it solves easy tasks from DMControl but have yet to try harder tasks + figure out what the right balance of envs and model updates are; `num_envs=4 steps_per_update=2` seems to be a good starting point.

**Comment by nicklashansen** [comment]
Benchmark results on `task=walker-walk`. Orange is default; green is vectorized (`num_envs = steps_per_update = 4`). Approx. 8 min vs. 24 min. to converge on a 2080-Ti.

<img width="392" alt="Screenshot 2024-02-11 at 4 57 46 PM" src="https://github.com/nicklashansen/tdmpc2/assets/38315247/5af811c2-8241-4059-bb1e-62e9e9f9e962">

**Comment by nicklashansen** [comment]
`task=humanoid-walk`, again default vs. `num_envs = steps_per_update = 4` on a 2080-Ti. Seems like it's also a ~3x speedup.

<img width="407" alt="Screenshot 2024-02-12 at 3 04 46 PM" src="https://github.com/nicklashansen/tdmpc2/assets/38315247/9b818c73-1a8b-4adb-913e-a6605289d28a">

**Comment by josephrocca** [comment]
Oh this looks great! I've just kicked off a few experiments including testing vectorized, playing with iterations, model size, adding in the merging on top of the vectorized env stuff, and some other random stuff. Will report back 🫡

**Comment by ShaneFlandermeyer** [comment]
Hi guys. The improvements in this thread are awesome to see! 

I am also interested in speeding up this algorithm, so I created a (mostly) 1:1 port in Jax/Flax. I am happy to report a 5-10x speedup for the planning/update steps, which reduces the training time to ~2.5 hours/million time steps! I imagine that we could get it down to an hour or less using multiple workers. If anyone is interested in trying to combine them, I have put my implementation in the repo below!

https://github.com/ShaneFlandermeyer/tdmpc2-jax

**Comment by josephrocca** [comment]
Wow, Shane! I should be able to get back to tdmpc experiments soon 🤞 can't wait to try out your code and test with the other ideas/improvements in this thread. What a time to be alive.

**Comment by nicklashansen** [comment]
Closing this issue. There has been recent efforts to speed up both training and inference of TD-MPC2, which has resulted in approx. 4.5x wall-time speedup compared to the original implementation. These changes have been pushed to `main` and are available with `compile=true` (requires nightly versions of `pytorch`, `tensordict`, `torchrl`). Our changes are orthogonal to environment vectorization, which (as evident from this thread) is an additional ~3x speedup. The `vectorized_env` branch of this repo includes both features.
